rasa/core/channels/twilio_voice.py

```diff
@@ -0,0 +1,335 @@
+            if (text is None) and (call_status == "ringing"):
+                text = self.initial_prompt
+
+            # determine the response.
+            if text is not None:
+                await on_new_message(
+                    UserMessage(
+                        text, collector, sender_id, input_channel=input_channel,
+                    )
+                )
+
+                twilio_response = self.build_twilio_voice_response(collector.messages)
+            # If the user doesn't respond resend the last message.
+            else:
+                # Get last user utterance from tracker.
+                tracker = request.app.agent.tracker_store.retrieve(sender_id)
+                last_response = next(
+                    (e for e in reversed(tracker.events) if isinstance(e, BotUttered)),
+                    None,
+                )
+
+                # If no previous utterance found use the reprompt_fallback phrase.
+                if last_response is None:
+                    last_response = self.reprompt_fallback_phrase
+                else:
+                    last_response = last_response.text
+
+                twilio_response = self.build_twilio_voice_response(
+                    [{"text": last_response}]
+                )
+            return response.text(str(twilio_response), content_type="text/xml")
+
+        return twilio_voice_webhook
+
+    def build_twilio_voice_response(
+        self, messages: List[Dict[Text, Any]]
+    ) -> VoiceResponse:
+        """Builds the Twilio Voice Response object."""
+        voice_response = VoiceResponse()
+        gather = Gather(
+            input="speech",
+            action=f"/webhooks/{self.name()}/webhook",
+            actionOnEmptyResult=True,
+            speechTimeout=self.speech_timeout,
+            speechModel=self.speech_model,
+            enhanced=self.enhanced,
+        )
+
+        # Add pauses between messages.
+        # Add a listener to the last message to listen for user response.
+        for i, message in enumerate(messages):
+            if i + 1 == len(messages):
+                gather.say(message["text"], voice=self.assistant_voice)
+                voice_response.append(gather)
+            else:
+                voice_response.say(message["text"], voice=self.assistant_voice)
+                voice_response.pause(length=1)
+
+        return voice_response
+
+
+class TwilioVoiceCollectingOutputChannel(CollectingOutputChannel):
+    """Output channel that collects send messages in a list.
+
+    (doesn't send them anywhere, just collects them).
+    """
+
+    @classmethod
+    def name(cls) -> Text:
+        """Name of the output channel."""
+        return "twilio_voice"
+
+    @staticmethod
+    def emoji_warning(
+        text: Text,
+    ) -> None:
+        """Raises a warning if text contains an emoji."""
+        tok = WhitespaceTokenizer()
+        emoji_regex = tok.get_emoji_regex()
```

Comment:
We don't really need the tokenizer - how about moving the `get_emoji` regex to something like `rasa.utils.io`?

Reply:
`get_emoji` was moved to `rasa.utils.io` and the reference was updated in both `twilio_voice.py` and in `WhitespaceTokenizer` to utilize the `get_emoji` utility.